### PR TITLE
fix(components): [date-picker] prevent duplicate pick date

### LINKED ISSUE #8112: [Bug Report] [Component] [date-picker] 日期选择上，如果选择的是当前月份的面板上的下个月的日期，比如六月份面板上选择七月份的三号，最后显示的是09-03
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.35`
- Element Plus Version: `2.2.4`
- Browser / OS: `macOS12.4`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[Docs CN](https://element-plus.gitee.io/zh-CN/component/date-picker.html)

### Steps to reproduce

1，在日期选择器的 上 打开选择面板
2，不切换月份，直接选择07-03

### What is Expected?

显示正确的选择日期

### What is actually happening?

实际上显示了下下个月（09-03）这个日期，好像点击事件触发了几次

### Additional comments

_(empty)_

### LINKED ISSUE #8489: [Component] [date-picker] 选择下一个月灰色的日期,返回日期与所选的不一样
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.37`
- Element Plus Version: `2.2.6`
- Browser / OS: `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgdmVyc2lvbiBhcyB2dWVWZXJzaW9uIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgdmVyc2lvbiBhcyBFcFZlcnNpb24gfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBFbGVtZW50UGx1cyB9IGZyb20gJ0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlJ1xuXG5cbmNvbnN0IHZhbHVlID0gcmVmKCcnKVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cblxuICA8cD5cbiAgICA8ZWwtZGF0ZS1waWNrZXJcbiAgICAgICAgdi1tb2RlbD1cInZhbHVlXCJcbiAgICAgICAgdHlwZT1cImRhdGVcIlxuICAgICAgICBwbGFjZWhvbGRlcj1cIlBpY2sgYSBkYXlcIiAgICAgXG4gICAgICAvPlxuICA8L3A+XG4gICBFbGVtZW50IFBsdXMge3sgRXBWZXJzaW9uIH19ICsgVnVlIHt7IHZ1ZVZlcnNpb24gfX1cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

1. Today is 2022-06-27, directly select the date 07-02 (just select the gray date)
2. will be return 2022-07-30 ,console error: Maximum call stack size exceeded will be reported.

### What is Expected?

Return the correct date selected

### What is actually happening?

return wrong date

### Additional comments

It will only appear if the initialized value is empty

### LINKED ISSUE #8494: [Component] [date-picker] DatePicker 在默认空值的时候 选择非本月日期会出现问题
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.37`
- Element Plus Version: `2.2.6`
- Browser / OS: `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36`
- Build Tool: `Other`

### Related Component

- `el-date-picker`

### Reproduction Link

[Docs CN](https://element-plus.gitee.io/zh-CN/component/datetime-picker.html#%E6%97%A5%E6%9C%9F%E5%92%8C%E6%97%B6%E9%97%B4%E7%82%B9)

### Steps to reproduce

官网下 demo
当值为空的时候 选择非当前月日会出现选择错误问题

     

### What is Expected?

当值为空的时候
当前是2022/6/27 选择非当前月的日期 
如选择5月31日应该还是5月31
如选择7月1日应该还是7月1

### What is actually happening?

当值为空的时候 
当前是2022/6/27 选择非当前月的日期 
如选择5月31日会变成2022-04-26
如选择7月1日会变成2022-7-29

### Additional comments

官网下 demo 直接可以尝试

https://user-images.githubusercontent.com/10862092/175909011-f1da2c9a-d3c7-495e-8886-b17531dd040b.mp4

### LINKED ISSUE #8498: [Component] [date-picker] Date Picker selects incorrect date sometimes
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.37`
- Element Plus Version: `2.2.6`
- Browser / OS: `Chrome/103.0.5060.53 / Windows 11`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgdmVyc2lvbiBhcyB2dWVWZXJzaW9uIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgdmVyc2lvbiBhcyBFcFZlcnNpb24gfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBFbGVtZW50UGx1cyB9IGZyb20gJ0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlJ1xuXG5jb25zdCBkYXRlID0gcmVmKG5ldyBEYXRlKCkpO1xuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPGgxPnt7IGRhdGUgfX08L2gxPlxuICA8ZWwtZGF0ZS1waWNrZXIgdi1tb2RlbD1cImRhdGVcIiAvPlxuXG4gIDxwPlxuICAgIDxlbC1pY29uIGNvbG9yPVwidmFyKC0tZWwtY29sb3ItcHJpbWFyeSlcIj48RWxlbWVudFBsdXMgLz48L2VsLWljb24+XG4gICAgRWxlbWVudCBQbHVzIHt7IEVwVmVyc2lvbiB9fSArIFZ1ZSB7eyB2dWVWZXJzaW9uIH19XG4gIDwvcD5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

1. Ensure the Date Picker already has a value selected
2. Open the Date Picker
3. Change the month (forward or backward) (e.g. from June to July)
4. Select a date for the previous or next month (e.g. June 30 or August 1)

### What is Expected?

The date that was clicked was selected

### What is actually happening?

A date approximately four weeks later or earlier is selected (e.g. Apr 25 or Sep 5)

### Additional comments

_(empty)_

### LINKED ISSUE #8520: [Component] [table-v2, date-picker] [Component][el-table-v2][el-date-picker]在el-table-v2 中使用  ElDatePicker 做自定义元素时无法获得选中的值  并且控制台会报错
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.33`
- Element Plus Version: `2.2.6`
- Browser / OS: `UserAgent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.0.0 Safari/537.36`
- Build Tool: `Vue CLI`

### Related Component

- `el-table-v2`
- `el-date-picker`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2IHN0eWxlPVwiaGVpZ2h0OiA0MDBweFwiPlxuICAgIDxlbC1hdXRvLXJlc2l6ZXI+XG4gICAgICA8dGVtcGxhdGUgI2RlZmF1bHQ9XCJ7IGhlaWdodCwgd2lkdGggfVwiPlxuICAgICAgICA8ZWwtdGFibGUtdjIgOmNvbHVtbnM9XCJjb2x1bW5zXCIgOmRhdGE9XCJkYXRhXCIgOndpZHRoPVwid2lkdGhcIiA6aGVpZ2h0PVwiaGVpZ2h0XCIgZml4ZWQgLz5cbiAgICAgIDwvdGVtcGxhdGU+XG4gICAgPC9lbC1hdXRvLXJlc2l6ZXI+XG4gIDwvZGl2PlxuXG4gIDxlbC1kYXRlLXBpY2tlciB2LW1vZGVsPVwidmFsdWVcIiAvPlxuPC90ZW1wbGF0ZT5cblxuPHNjcmlwdCA+XG5pbXBvcnQgeyByZWFjdGl2ZSwgcmVmLCB0b1JlZnMsIG9uTW91bnRlZCwgaCB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB7IEVsRGF0ZVBpY2tlcixFbElucHV0IH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuXG5leHBvcnQgZGVmYXVsdCB7XG4gIG5hbWU6ICd0YWJsZVYyJyxcbiAgc2V0dXAoKSB7XG4gICAgb25Nb3VudGVkKCgpID0+IHt9KVxuXG4gICAgbGV0IHRhYmxlRGF0YSA9IHJlYWN0aXZlKHtcbiAgICAgIGRhdGE6IFtcbiAgICAgICAge1xuICAgICAgICAgIHRleHQ6ICdxaWV1cm9xdXdlb3InLFxuICAgICAgICAgIHRleHRFZGl0OiBmYWxzZSxcbiAgICAgICAgICBuYW1lOiAnMTIzMTIzJyxcbiAgICAgICAgICBuYW1lRWRpdDogZmFsc2UsXG4gICAgICAgICAgYWdlOiAwLFxuICAgICAgICAgIGFnZUVkaXQ6IGZhbHNlLFxuICAgICAgICAgIHNleDogJycsXG4gICAgICAgICAgc2V4RWRpdDogZmFsc2UsXG4gICAgICAgICAgbGlrZTogW10sXG4gICAgICAgICAgbGlrZUVkaXQ6IGZhbHNlLFxuICAgICAgICAgIHdvcms6ICcnLFxuICAgICAgICAgIHdvcmtFZGl0OiBmYWxzZSxcbiAgICAgICAgICB1c2VyTmFtZTogJycsXG4gICAgICAgICAgdXNlck5hbWVFZGl0OiBmYWxzZSxcbiAgICAgICAgICBwYXNzd29yZDogJycsXG4gICAgICAgICAgcGFzc3dvcmRFZGl0OiBmYWxzZSxcbiAgICAgICAgICBiaXJ0aGRheTogJycsXG4gICAgICAgICAgYmlydGhkYXlFZGl0OiBmYWxzZSxcbiAgICAgICAgICBjcmVhdGVUaW1lOiAnJyxcbiAgICAgICAgICBjcmVhdGVUaW1lRWRpdDogZmFsc2UsXG4gICAgICAgICAgbGVhdmVUaW1lOiAnJyxcbiAgICAgICAgICBsZWF2ZVRpbWVFZGl0OiBmYWxzZVxuICAgICAgICB9XG4gICAgICBdLFxuICAgICAgY29sdW1uc1R5cGVPYmplY3Q6IHtcbiAgICAgICAgdGV4dDoge1xuICAgICAgICAgIHRpdGxlOiAn5paH5pysJyxcbiAgICAgICAgICB0eXBlOiAnVGV4dCcsXG4gICAgICAgICAgd2lkdGg6IDIwMFxuICAgICAgICB9LFxuICAgICAgICBuYW1lOiB7XG4gICAgICAgICAgdGl0bGU6ICflkI3np7AnLFxuICAgICAgICAgIHR5cGU6ICdJbnB1dCcsXG4gICAgICAgICAgd2lkdGg6IDQwMFxuICAgICAgICB9LFxuICAgICAgICBiaXJ0aGRheToge1xuICAgICAgICAgIHRpdGxlOiAn55Sf5pelJyxcbiAgICAgICAgICB0eXBlOiAnRGF0ZVBpY2tlcidcbiAgICAgICAgfSxcbiAgICAgIH0sXG4gICAgICBjb2x1bW5zOiBbXSxcbiAgICAgIGluaXQoKSB7XG4gICAgICAgIGxldCBhcnIgPSBPYmplY3Qua2V5cyh0YWJsZURhdGEuY29sdW1uc1R5cGVPYmplY3QpXG4gICAgICAgIGxldCBjb2x1bW5zID0gW11cbiAgICAgICAgYXJyLmZvckVhY2goKGtleSwgaW5kZXgpID0+IHtcbiAgICAgICAgICBsZXQgaXRlbSA9IHRhYmxlRGF0YS5jb2x1bW5zVHlwZU9iamVjdFtrZXldXG4gICAgICAgICAgbGV0IGRlZmF1bHRSb3dJdGVtID0ge1xuICAgICAgICAgICAga2V5OiAnY29sdW1uLScgKyBpbmRleCxcbiAgICAgICAgICAgIGRhdGFLZXk6IGtleSxcbiAgICAgICAgICAgIHRpdGxlOiBpdGVtLnRpdGxlLFxuICAgICAgICAgICAgd2lkdGg6IGl0ZW0ud2lkdGggfHwgMjAwXG4gICAgICAgICAgfVxuXG4gICAgICAgICAgY29sdW1uc1tpbmRleF0gPSBkZWZhdWx0Um93SXRlbVxuICAgICAgICAgIHN3aXRjaCAoaXRlbS50eXBlKSB7XG4gICAgICAgICAgICBjYXNlICdUZXh0JzpcbiAgICAgICAgICAgICAgY29sdW1uc1tpbmRleF0gPSBkZWZhdWx0Um93SXRlbVxuICAgICAgICAgICAgICBicmVha1xuICAgICAgICAgICAgY2FzZSAnSW5wdXQnOlxuICAgICAgICAgICAgICAgICBjb2x1bW5zW2luZGV4XVsnY2VsbFJlbmRlcmVyJ10gPSAoeyByb3dEYXRhLCBjb2x1bW4gfSkgPT5cbiAgICAgICAgICAgICAgICBoKEVsSW5wdXQsIHtcbiAgICAgICAgICAgICAgICAgIG1vZGVsVmFsdWU6IHJvd0RhdGFbY29sdW1uLmRhdGFLZXldLFxuICAgICAgICAgICAgICAgIH0pXG4gICAgICAgICAgICAgIGJyZWFrXG4gICAgICAgICAgIGNhc2UgJ0RhdGVQaWNrZXInOlxuICAgICAgICAgICAgICBjb2x1bW5zW2luZGV4XVsnY2VsbFJlbmRlcmVyJ10gPSAoeyByb3dEYXRhLCBjb2x1bW4gfSkgPT5cbiAgICAgICAgICAgICAgICBoKEVsRGF0ZVBpY2tlciwge1xuICAgICAgICAgICAgICAgICAgbW9kZWxWYWx1ZTogcm93RGF0YVtjb2x1bW4uZGF0YUtleV1cbiAgICAgICAgICAgICAgICB9KVxuICAgICAgICAgICAgICBicmVha1xuICAgICAgICAgIH1cblxuICAgICAgICAgIGlmIChpbmRleCA9PT0gYXJyLmxlbmd0aCAtIDEpIHtcbiAgICAgICAgICAgIHRhYmxlRGF0YS5jb2x1bW5zID0gY29sdW1uc1xuICAgICAgICAgICAgY29uc29sZS5sb2coY29sdW1ucylcbiAgICAgICAgICB9XG4gICAgICAgIH0pXG4gICAgICB9XG4gICAgfSlcbiAgICB0YWJsZURhdGEuaW5pdCgpXG5cbiAgICBsZXQgdmFsdWUgPSByZWYoJycpXG4gICAgcmV0dXJuIHtcbiAgICAgIC4uLnRvUmVmcyh0YWJsZURhdGEpLFxuICAgICAgdmFsdWVcbiAgICB9XG4gIH1cbn1cbjwvc2NyaXB0PlxuXG48c3R5bGU+XG4udGFibGUtdjItaW5saW5lLWVkaXRpbmctdHJpZ2dlciB7XG4gIGJvcmRlcjogMXB4IHRyYW5zcGFyZW50IGRvdHRlZDtcbiAgd2lkdGg6IDEwMCU7XG4gIGxpbmUtaGVpZ2h0OiAzMnB4O1xuICBwYWRkaW5nOiAwIDEwcHg7XG4gIG1pbi1oZWlnaHQ6IDMycHg7XG59XG5cbi50YWJsZS12Mi1pbmxpbmUtZWRpdGluZy10cmlnZ2VyOmhvdmVyIHtcbiAgYm9yZGVyLWNvbG9yOiB2YXIoLS1lbC1jb2xvci1wcmltYXJ5KTtcbn1cbjwvc3R5bGU+IiwiaW1wb3J0X21hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge31cbn0iLCJfbyI6e319)

### Steps to reproduce

在el-table-v2 中使用自定义元素 [el-date-picker] 可以正常显示元素  ==> 但是操作选择时间后控制台会报错 

报错信息
```
_objectToString.js:3 Uncaught RangeError: Maximum call stack size exceeded.
    at W (_objectToString.js:3:32)
    at callWithErrorHandling (runtime-dom.esm-browser.js:1651:22)
    at callWithAsyncErrorHandling (runtime-dom.esm-browser.js:1660:21)
    at emit$1 (runtime-dom.esm-browser.js:2210:9)
    at runtime-dom.esm-browser.js:8724:53
    at HTMLDivElement.c (_objectToString.js:3:32)
    at W (_objectToString.js:3:32)
    at callWithErrorHandling (runtime-dom.esm-browser.js:1651:22)
    at callWithAsyncErrorHandling (runtime-dom.esm-browser.js:1660:21)
    at emit$1 (runtime-dom.esm-browser.js:2210:9)
```

### What is Expected?

期望 el-date-picker时间选择器能在 el-table-v2虚拟表格中正确返回选择时间

### What is actually happening?

```
_objectToString.js:3 Uncaught RangeError: Maximum call stack size exceeded.
    at W (_objectToString.js:3:32)
    at callWithErrorHandling (runtime-dom.esm-browser.js:1651:22)
    at callWithAsyncErrorHandling (runtime-dom.esm-browser.js:1660:21)
    at emit$1 (runtime-dom.esm-browser.js:2210:9)
    at runtime-dom.esm-browser.js:8724:53
    at HTMLDivElement.c (_objectToString.js:3:32)
    at W (_objectToString.js:3:32)
    at callWithErrorHandling (runtime-dom.esm-browser.js:1651:22)
    at callWithAsyncErrorHandling (runtime-dom.esm-browser.js:1660:21)
    at emit$1 (runtime-dom.esm-browser.js:2210:9)
```

### Additional comments

_(empty)_

### LINKED ISSUE #8523: [Component] [date-picker] [Component][el-date-picker]使用ElDatePicker没有Onchange事件以及会发生控制台异常
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.26`
- Element Plus Version: `2.2.6`
- Browser / OS: `Chrome`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgdmVyc2lvbiBhcyB2dWVWZXJzaW9uIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgdmVyc2lvbiBhcyBFcFZlcnNpb24gfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBFbGVtZW50UGx1cyB9IGZyb20gJ0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlJ1xuXG5jb25zdCBtc2cgPSByZWYoJ0hlbGxvIFdvcmxkIScpXG5jb25zdCBtc2cxID0gcmVmKCcnKVxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPGgxPnt7IG1zZyB9fTwvaDE+XG4gIDxlbC1pbnB1dCB2LW1vZGVsPVwibXNnXCIgLz5cblxuICA8cD48RWxEYXRlUGlja2VyIG1vZGVsVmFsdWU9e21zZzF9IG9uQ2hhbmdlPXsodmFsdWUpPT57Y29uc29sZS5sb2codmFsdWUpfX0vPjwvcD5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

打开playground

### What is Expected?

正常运行

onchange事件打印

### What is actually happening?

控制台提示：
_objectToString.js:3 Uncaught RangeError: Maximum call stack size exceeded.
onchange事件未打印

### Additional comments

_(empty)_

### LINKED ISSUE #8545: [Component] [date-picker] 日期选择器
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.37`
- Element Plus Version: `2.2.6`
- Browser / OS: `Microsoft Edge 版本 103.0.1264.37 (正式版本) (64 位)`
- Build Tool: `Webpack`

### Related Component

- `el-date-picker`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgdmVyc2lvbiBhcyB2dWVWZXJzaW9uIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgdmVyc2lvbiBhcyBFcFZlcnNpb24gfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBFbGVtZW50UGx1cyB9IGZyb20gJ0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlJ1xuY29uc3QgZGF0ZSA9IHJlZignJylcbmNvbnN0IG1zZyA9IHJlZignSGVsbG8gV29ybGQhJylcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxoMT57eyBtc2cgfX08L2gxPlxuICA8ZWwtaW5wdXQgdi1tb2RlbD1cIm1zZ1wiIC8+XG5cdDxlbC1kYXRlLXBpY2tlclxuICAgICAgICAgIHYtbW9kZWw9XCJkYXRlXCJcbiAgICAgICAgICAgICAgICAgdHlwZT1cImRhdGVcIlxuICAgICAgICAgICAgICAgICAgICBwbGFjZWhvbGRlcj1cIuivt+mAieaLqeaXtumXtFwiXG4gICAgICAgICAgICAgICAgICAgIHZhbHVlRm9ybWF0PVwiWVlZWS1NTS1ERFwiXG4gICAgICAgICAgICAgICAgICAgIC8+XG4gIDxwPlxuICAgIDxlbC1pY29uIGNvbG9yPVwidmFyKC0tZWwtY29sb3ItcHJpbWFyeSlcIj48RWxlbWVudFBsdXMgLz48L2VsLWljb24+XG4gICAgRWxlbWVudCBQbHVzIHt7IEVwVmVyc2lvbiB9fSArIFZ1ZSB7eyB2dWVWZXJzaW9uIH19XG4gIDwvcD5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

日期选择器，今天是6月30，现在选择当前界面的7月1日，会跳转到7月29,会报错

### What is Expected?

选择为7月1日

### What is actually happening?

7月29

### Additional comments

_(empty)_

### LINKED ISSUE #8559: [Component] [date-picker] 日期组件bug,日期下拉框在不切换月份时，第一次选中灰色下个月任意日期时，日期对应不上，选中框中显示的是这个灰色月末日期，切换下月份在选择后又正常了
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.37`
- Element Plus Version: `2.2.5`
- Browser / OS: `chrome`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[Docs CN](https://element-plus.gitee.io/zh-CN/component/date-picker.html#选择某一天)

### Steps to reproduce

日期下拉框在不切换月份时，第一次选中灰色下个月任意日期时，日期对应不上，选中框中显示的是这个灰色月末日期（比如我选中灰色下个月1号选中框中显示30号），切换下月份在选择后又正常了

### What is Expected?

第一次选中灰色显示对应日期

### What is actually happening?

第一次选中灰色显示对应日期

### Additional comments

_(empty)_

### LINKED ISSUE #8728: [Component] [date-picker] Uncaught RangeError: Maximum call stack size exceeded.
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.37`
- Element Plus Version: `2.2.8`
- Browser / OS: `UserAgent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdD5cbmltcG9ydCB7IHJlZiwgaCB9IGZyb20gJ3Z1ZSdcbmltcG9ydCB7IEVsZW1lbnRQbHVzIH0gZnJvbSAnQGVsZW1lbnQtcGx1cy9pY29ucy12dWUnXG5pbXBvcnQgeyAgRWxEYXRlUGlja2VyIH0gZnJvbSAnZWxlbWVudC1wbHVzJ1xuZXhwb3J0IGRlZmF1bHQge1xuICBzZXR1cCgpIHtcblx0XHRjb25zdCBfZGF0ZVZhbHVlID0gcmVmKCcnKVxuICAgIHJldHVybiAoKSA9PiB7XG4gICAgICByZXR1cm4gaChcbiAgICAgICAgICAgIEVsRGF0ZVBpY2tlcixcbiAgICAgICAgICAgIHtcbiAgICAgICAgICAgICAgbW9kZWxWYWx1ZTogX2RhdGVWYWx1ZS52YWx1ZSxcbiAgICAgICAgICAgICAgdHlwZTogJ2RhdGUnLFxuICAgICAgICAgICAgICBwbGFjZWhvbGRlcjogJ+ivt+mAieaLqScsXG4gICAgICAgICAgICB9XG4gICAgICAgICAgKVxuICAgIH1cbiAgfVxufVxuXG48L3NjcmlwdD5cbiIsImltcG9ydF9tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHt9XG59IiwiX28iOnt9fQ==)

### Steps to reproduce

Date cannot be selected

### What is Expected?

Can choose a date

### What is actually happening?

Date cannot be selected

### Additional comments

_(empty)_

### LINKED ISSUE #8802: [Component] [date-picker] DatePicker组件首次进入显示日期与选择日期不一致，当前日期中选择上月日期，会触发加载上月日期并选择同位置的上月日期，如在7月的日历上选择2022-06-31，显示为2022-06-02
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.31`
- Element Plus Version: `2.0.4`
- Browser / OS: `Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[Docs CN](https://element-plus.gitee.io/zh-CN/component/date-picker.html#%E8%87%AA%E5%AE%9A%E4%B9%89%E5%86%85%E5%AE%B9)

### Steps to reproduce

1、打开日期选择器
2、选择上一月的任意日期
3、查看选中日期

### What is Expected?

与点击日期保持一致

### What is actually happening?

不一致，选中了上一月的同一位置的日期【第二次之后正常】

### Additional comments

_(empty)_

### LINKED ISSUE #8840: [Component] [date-picker] [组件] [日期选择器] el-date-picker 当前月首次选择下个月，会增加一个月的时间
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.37`
- Element Plus Version: `2.2.9`
- Browser / OS: `Chrome 103.0.5060.114`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[Docs](https://element-plus.org/zh-CN/component/date-picker.html)

### Steps to reproduce

当选择首次选择下一个月时实际日期增加了一个月，并报错：Uncaught RangeError: Maximum call stack size exceeded

### What is Expected?

在当前月能够正常选择下一个的日期

### What is actually happening?

在当前月首次选择下一个月日期时，实际显示的日期会增加一个月
例子：6月窗口选择7月1日，实际显示为9月1日

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by holazz**
- fix duplicate pick date
- fix the keyboard selection not working when opening the picker panel for the first time

closes #8559, closes #8545, closes #8523, closes #8520, closes #8498, closes #8494, closes #8489, closes #8112, closes #8728, closes #8802, closes #8840

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by holazz** [comment]
> Maybe use `mousedown` instead of `click`
> 
> mousedown -> focus -> mouseup -> click

Do you mean to prevent the default event on mousedown to prevent the focus event from being fired?

`@click="handlePickDate"` -> `@mousedown.prevent="handlePickDate"`

**Comment by holazz** [comment]
> `click` -> `mousedown`

Use click, if you don't want to select after pressing the mouse, you can cancel the pick date action, like this:

https://user-images.githubusercontent.com/23100055/179329157-b8eb9691-d7f3-4556-8d7a-5a67a98e0401.mov

Use mousedown, the pick date action is triggered the moment you press the mouse.


I'm not sure which behavior is correct. WDYT?

**Comment by holazz** [comment]
I have now replaced `click` with `mousedown`.

**Comment by tolking** [comment]
> Use mousedown, the pick date action is triggered the moment you press the mouse.
> 
> I'm not sure which behavior is correct. WDYT?

Sorry for not taking this into consideration before, I looked at the built-in date input component, and it is really not very appropriate to use mousedown events.

```html
<input type="date" />
```

**Comment by holazz** [comment]
> > Use mousedown, the pick date action is triggered the moment you press the mouse.
> > I'm not sure which behavior is correct. WDYT?
> 
> Sorry for not taking this into consideration before, I looked at the built-in date input component, and it is really not very appropriate to use mousedown events.
> 
> ```
> <input type="date" />
> ```

Ok, I'll change it back to click later.

**Comment by holazz** [comment]
@tolking PTAL

File: packages/components/date-picker/src/date-picker-com/basic-date-table.vue [hunk 1]
```diff
@@ -68,6 +70,8 @@ const lastRow = ref<number>()
 const lastColumn = ref<number>()
 const tableRows = ref<DateCell[][]>([[], [], [], [], [], []])
 
+const focusWithClick = ref(false)
```
Comment (tolking) [inline packages/components/date-picker/src/date-picker-com/basic-date-table.vue:73]: Maybe there is no need to use `ref` here

```suggestion
let focusWithClick = false
```

Comment (holazz) [reply packages/components/date-picker/src/date-picker-com/basic-date-table.vue:73]: you're right. 👍

**Review by tolking — APPROVED** [review]

**Pull request merged**